### PR TITLE
fix: use regex in iwyu.imp

### PR DESCRIPTION
**Pull request by wdconinc**
### Briefly, what does this PR introduce?
Instead of exhaustive lists in iwyu.imp we can use regex now. Avoids overlooking additions like we had with the measurements datatype.

TODO:
- [x] revert https://github.com/eic/EICrecon/pull/1162/commits/00d9f3d36b6ce61c5d8182110c8445813f9dacec
- [x] revert https://github.com/eic/EICrecon/pull/1162/commits/26bea2436ae8ad40dfa6537c6acd86ac4a024a12

### What kind of change does this PR introduce?
- [x] Bug fix (issue #__)
- [ ] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No.

### Does this PR change default behavior?
No.

### CODE REVIEW

**Comment by wdconinc** [comment]
The problem with my regex is that it still results in additions like this:
```diff
+#include <edm4eic/MCRecoParticleAssociation.h>
 #include <edm4eic/MCRecoParticleAssociationCollection.h>
```
That's what I get with the following regex:
```
#  { include: ['@<edm4(eic|hep)/(.*)Data\.h>', private, '<edm4\1/\2Collection.h>', public] },
#  { include: ['@<edm4(eic|hep)/Mutable(.*)\.h>', private, '<edm4\1/\2Collection.h>', public] },
  { include: ['@<edm4(eic|hep)/(Mutable)?([^/]*?)(Collection|Data|Obj)?\.h>', private, '<edm4\1/\2Collection.h>', public] },
```
which seems like it should work based on https://regex101.com/r/jFr6NE/1 (small changes there).

**Comment by wdconinc** [comment]
Grr, why claim regex=ecmascript but then only support part of ecmascript. https://en.cppreference.com/w/cpp/regex/ecmascript is what is used. Doesn't support named capture groups.

**Comment by wdconinc** [comment]
It is really annoying that the only sign of an unparsable regex is that nothing happens.

**Comment by wdconinc** [comment]
Regex in iwyu.imp now ok.

**Comment by wdconinc** [comment]
Thanks IWYU, this is unstable. Add Householder, remove Householder. I'm not playing that game.

File: .github/iwyu.imp [hunk 1]
```diff
@@ -2,249 +2,31 @@
...
-  { include: ["<edm4hep/TrackData.h>", "private", "<edm4hep/TrackCollection.h>", "public"] },
-  { include: ["<edm4hep/TrackerHitData.h>", "private", "<edm4hep/TrackerHitCollection.h>", "public"] },
-  { include: ["<edm4hep/TrackerHitPlaneData.h>", "private", "<edm4hep/TrackerHitPlaneCollection.h>", "public"] },
-  { include: ["<edm4hep/TrackerPulseData.h>", "private", "<edm4hep/TrackerPulseCollection.h>", "public"] },
-  { include: ["<edm4hep/VertexData.h>", "private", "<edm4hep/VertexCollection.h>", "public"] },
+  # Eigen
+  { include: ['@<Eigen/(src/)?(.*?)/.*>', private, '<Eigen/$2>', public] },
+  { include: ['@<Eigen/src/plugins/.*>', private, '<Eigen/Core>', public] },
+  { include: ['@"src/(.*)/.*"', private, '<Eigen/$1>', public] },
+  { include: ['@"src/plugins/.*"', private, '<Eigen/Core>', public] },
```
[the remark this answers is not in the document]
Comment (veprbl) [reply .github/iwyu.imp:21]: Apparently it also does some correct things:
https://gitlab.com/libeigen/eigen/-/merge_requests/1199
Including export pragma for Eigen/Core and private pragma for some src/plugins includes
https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/ArrayBase.h#L13

[the remark this answers is not in the document]
Comment (wdconinc) [reply .github/iwyu.imp:21]: Does iwyu 0.20 have support for those pragmas?

[the remark this answers is not in the document]
Comment (wdconinc) [reply .github/iwyu.imp:21]: Ah, it's in eigen main branch but not in a release yet.

**Pull request merged**